pysmt/test/test_bv_simplification.py

```diff
@@ -94,6 +95,316 @@ def all_special(self):
                 for op in self.special_operators:
                     self.check(op(l, n))
 
+    def test_bv_0_add(self):
+        x = Symbol("x", BVType(32))
+        f = BVAdd(BVZero(32), x)
+        self.assertEqual(f.simplify(), Symbol("x", BVType(32)))
+
+    def test_bv_add_0(self):
+        x = Symbol("x", BVType(32))
+        f = BVAdd(x, BVZero(32))
+        self.assertEqual(f.simplify(), Symbol("x", BVType(32)))
+
+    def test_bv_add_constants(self):
+        f = BVAdd(BV(10, 32), BV(12, 32))
+        self.assertEqual(f.simplify(), BV(22, 32))
+
+    def test_bv_add_symbols(self):
+        x, y = (Symbol(name, BVType(32)) for name in "xy")
+        f = BVAdd(x, y)
+        self.assertEqual(f.simplify(), BVAdd(x, y))
+
+    def test_bv_0_mul(self):
+        x = Symbol("x", BVType(32))
+        f = BVMul(BVZero(32), x)
+        self.assertEqual(f.simplify(), BVZero(32))
+
+    def test_bv_1_mul(self):
+        x = Symbol("x", BVType(32))
+        f = BVMul(BVOne(32), x)
+        self.assertEqual(f.simplify(), Symbol("x", BVType(32)))
+
+    def test_bv_mul_0(self):
+        x = Symbol("x", BVType(32))
+        f = BVMul(x, BVZero(32))
+        self.assertEqual(f.simplify(), BVZero(32))
+
+    def test_bv_mul_1(self):
+        x = Symbol("x", BVType(32))
+        f = BVMul(x, BVOne(32))
+        self.assertEqual(f.simplify(), Symbol("x", BVType(32)))
+
+    def test_bv_mul_constants(self):
+        f = BVMul(BV(10, 32), BV(12, 32))
+        self.assertEqual(f.simplify(), BV(120, 32))
+
+    def test_bv_mul_symbols(self):
+        x, y = (Symbol(name, BVType(32)) for name in "xy")
+        f = BVMul(x, y)
+        self.assertEqual(f.simplify(), BVMul(x, y))
+
+    def test_bv_udiv_1(self):
+        x = Symbol("x", BVType(32))
+        f = BVUDiv(x, BVOne(32))
+        self.assertEqual(f.simplify(), Symbol("x", BVType(32)))
+
+    def test_bv_symbol_udiv_0(self):
+        x = Symbol("x", BVType(32))
+        f = BVUDiv(x, BVZero(32))
+        self.assertEqual(f.simplify(), BV(2**32 - 1, 32))
+
+    def test_bv_0_udiv_0(self):
+        f = BVUDiv(BVZero(32), BVZero(32))
+        self.assertEqual(f.simplify(), BV(2**32 - 1, 32))
+
+    def test_bv_nonzero_udiv_0(self):
+        f = BVUDiv(BV(10, 32), BVZero(32))
+        self.assertEqual(f.simplify(), BV(2**32 - 1, 32))
```

Comment:
I don't remember what is the agreed value for this. This is undefined in arithmetic theories. Is this defined in the smtlib standard?

Reply:
Simplifier::walk_bv_udiv() before this PR does so, as well as Z3, though the bit vector division by zero seems invalid in [smtlib standard](http://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml):

>   o _ div _,  which takes an integer x ≥ 0 and an integer y > 0 and returns
>     the integer part of x divided by y (i.e., truncated integer division).
>
>   o _ rem _, which takes an integer x ≥ 0 and y > 0 and returns the
>     remainder when x is divided by y.  Note that we always have the following
>     equivalence for y > 0: (x div y) * y + (x rem y) = x.